File: src/day02/main.rs
use regex::Regex;
use shared::puzzle_input;

fn color_max(rounds: &str, color_name: &str) -> usize {
    let re = Regex::new(&format!(r"(\d+) {color_name}")).unwrap();
    rounds
        .split(';')
        .flat_map(|s| re.captures(s))
        .map(|caps| caps.get(1).unwrap().as_str().parse().unwrap())
        .max()
        .unwrap_or(0)
}

fn p1(line: &str) -> Option<usize> {
    let re = Regex::new(r"Game (\d+): (((\d+) (\w+))([^\d]*(\d+) (\w+))*)").ok()?;
    let caps = re.captures(line)?;
    let game_id: usize = caps.get(1)?.as_str().parse().ok()?;
    let rounds = caps.get(2)?.as_str();
    if color_max(rounds, "red") > 12 {
        return None;
    }
    if color_max(rounds, "green") > 13 {
        return None;
    }
    if color_max(rounds, "blue") > 14 {
        return None;
    }
    Some(game_id)
}

fn p2(line: &str) -> Option<usize> {
    let re = Regex::new(r"[^:]: (((\d+) (\w+))([^\d]*(\d+) (\w+))*)").ok()?;
    let caps = re.captures(line)?;
    let rounds = caps.get(1)?.as_str();
    Some(color_max(rounds, "red") * color_max(rounds, "green") * color_max(rounds, "blue"))
}

fn main() {
    let input = puzzle_input!();
    println!("{}", input.split('\n').flat_map(p1).sum::<usize>());
    println!("{}", input.split('\n').flat_map(p2).sum::<usize>());
}

File: src/day02/hs/main.hs
-- ghc .\src\day02\hs\main.hs ; .\src\day02\hs\main
import Data.Char (digitToInt, isDigit)
import Data.List (isPrefixOf)
import Maybes (mapMaybe)

neededCubes s = neededCubes' (0, 0, 0) s
  where
    neededCubes' cubes "" = cubes
    neededCubes' (r, g, b) s = neededCubes' (r', g', b') s''
      where
        (numString, s') = break (== ' ') $ dropWhile (not . isDigit) s
        num = read numString
        update (r, g, b) s
          | "red" `isPrefixOf` s = ((max r num, g, b), drop 3 s)
          | "green" `isPrefixOf` s = ((r, max g num, b), drop 5 s)
          | "blue" `isPrefixOf` s = ((r, g, max b num), drop 4 s)
          | otherwise = error $ "parse error: \"" ++ s ++ "\""
        ((r', g', b'), s'') = update (r, g, b) $ drop 1 s'

p1 lines = sum $ mapMaybe score_line lines
  where
    score_line line
      | "Game " `isPrefixOf` line = score_line $ drop 5 line
      | otherwise = do
        let (id, rest) = break (== ':') line
        let (r, g, b) = neededCubes $ drop 1 rest
        if (r > 12) || (g > 13) || (b > 14)
          then Nothing
          else Just $ read id

p2 lines = sum $ map scoreLine lines
  where
    scoreLine line = r * g * b
      where
        (r, g, b) = neededCubes $ drop 2 $ dropWhile (/= ':') line

main :: IO ()
main = do
  input <- readFile "puzzle_input/day02"
  print (p1 $ lines input) -- 2156
  print (p2 $ lines input) -- 66909